### PR TITLE
feat(bottlecap): add invocation span

### PR DESCRIPTION
**Pull request by duncanista**
# What?

Adds the `aws.lambda` invocation span to Universally Instrumented runtimes like .NET, Go, Java, and Ruby.

image

# Motivation

#374 and [SVLS-5438](https://datadoghq.atlassian.net/browse/SVLS-5438)

# Testing

- Added some unit tests, integration tests will be added later
- Tested manually for .NET

# Notes

Some tags will be missing from the top-level span

[SVLS-5438]: https://datadoghq.atlassian.net/browse/SVLS-5438?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: bottlecap/src/metrics/enhanced/constants.rs [hunk 1]
```diff
@@ -3,7 +3,8 @@ pub const BASE_LAMBDA_INVOCATION_PRICE: f64 = 0.000_000_2;
 pub const X86_LAMBDA_PRICE_PER_GB_SECOND: f64 = 0.000_016_666_7;
 pub const ARM_LAMBDA_PRICE_PER_GB_SECOND: f64 = 0.000_013_333_4;
 pub const MS_TO_SEC: f64 = 0.001;
-pub const MB_TO_GB: f64 = 1024.0;
+pub const MS_TO_NS: f64 = 1_000_000.0;
```
Comment (astuyve) [inline bottlecap/src/metrics/enhanced/constants.rs:6]: Do we need this for enhanced metrics?

File: bottlecap/src/lifecycle/invocation/context.rs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            request_id,
+            runtime_duration_ms,
+            init_duration_ms,
+            start_time,
+        }
+    }
+}
+
+#[allow(clippy::module_name_repetitions)]
+pub struct ContextBuffer {
```
Comment (duncanista) [inline bottlecap/src/lifecycle/invocation/context.rs:31]: I decided to use the `ContextBuffer` so we can keep relevant data from the past `5` invocations in cache.

5 is just an arbitrary number, we can reduce/increase this number as desired.

File: bottlecap/src/metrics/enhanced/constants.rs (hunk 1 above)
Comment (duncanista) [reply bottlecap/src/metrics/enhanced/constants.rs:6]: Not really, it's more in general, I might move this to the file where it's being used

Comment (astuyve) [reply bottlecap/src/metrics/enhanced/constants.rs:6]: sg

File: bottlecap/src/bin/bottlecap/main.rs [hunk 3]
```diff
@@ -375,92 +381,97 @@ async fn extension_loop_active(
...
-                                request_id,
-                                status,
-                                metrics,
-                                ..
-                            } => {
-                                if let Some(metrics) = metrics {
-                                    invocation_context_buffer
-                                        .add_runtime_duration(&request_id, metrics.duration_ms);
+                        Event::Telemetry(event) => {
+                            let event_time = event.time;
```
Comment (duncanista) [inline bottlecap/src/bin/bottlecap/main.rs:385]: maybe I can refactor this `event_time` to be passed directly, it's way too hard to see what changed here

File: bottlecap/src/lifecycle/invocation/processor.rs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            let header_tags = tracer_header_tags::TracerHeaderTags {
+                lang: "",
+                lang_version: "",
+                lang_interpreter: "",
+                lang_vendor: "",
+                tracer_version: "",
+                container_id: "",
+                client_computed_top_level: false,
+                client_computed_stats: false,
+            };
```
Comment (duncanista) [inline bottlecap/src/lifecycle/invocation/processor.rs:101]: definitely missing a lot of data here

File: bottlecap/src/lifecycle/invocation/processor.rs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        self.context_buffer
+            .add_runtime_duration(request_id, duration_ms);
+
+        if let Some(context) = self.context_buffer.get(&self.current_request_id) {
+            let span = &mut self.span;
+            // `round` is intentionally meant to be a whole integer
+            span.duration = (context.runtime_duration_ms * MS_TO_NS).round() as i64;
+            span.meta
+                .insert("request_id".to_string(), request_id.clone());
+        }
```
Comment (duncanista) [inline bottlecap/src/lifecycle/invocation/processor.rs:88]: Missing a bunch of other tags here too

File: bottlecap/src/bin/bottlecap/main.rs [hunk 6]
```diff
@@ -298,23 +302,26 @@ async fn extension_loop_active(
     let trace_flusher_clone = trace_flusher.clone();
     let stats_flusher_clone = stats_flusher.clone();
 
-    let trace_agent = Box::new(trace_agent::TraceAgent {
```
Comment (duncanista) [inline bottlecap/src/bin/bottlecap/main.rs:301]: Mmmm, I'm not sure if it was wrong to the remove the Box here

Comment (duncanista) [reply bottlecap/src/bin/bottlecap/main.rs:301]: Added it back

File: bottlecap/src/lifecycle/invocation/processor.rs (hunk 4 above)
Comment (duncanista) [reply bottlecap/src/lifecycle/invocation/processor.rs]: Added a todo comment, not sure how we can get this ones easily

File: bottlecap/src/bin/bottlecap/main.rs [hunk 7]
```diff
@@ -375,92 +383,95 @@ async fn extension_loop_active(
...
+                                            &request_id,
+                                            metrics.duration_ms,
+                                            config.clone(),
+                                            tags_provider.clone(),
+                                            trace_processor.clone(),
+                                            trace_agent_tx.clone()
+                                        ).await;
+                                        lambda_enhanced_metrics
+                                            .set_runtime_duration_metric(metrics.duration_ms);
+                                    }
```
Comment (duncanista) [inline bottlecap/src/bin/bottlecap/main.rs:420]: I might want to delegate some of this metrics to the lifecycle processor eventually, maybe?

File: bottlecap/src/bin/bottlecap/main.rs [hunk 8]
```diff
@@ -375,92 +383,95 @@ async fn extension_loop_active(
...
-                                metrics,
-                                ..
-                            } => {
-                                if let Some(metrics) = metrics {
-                                    invocation_context_buffer
-                                        .add_runtime_duration(&request_id, metrics.duration_ms);
+                        Event::Telemetry(event) =>
+                            match event.record {
+                                TelemetryRecord::PlatformStart { request_id, .. } => {
+                                    let mut p = invocation_processor.lock().await;
```
Comment (astuyve) [inline bottlecap/src/bin/bottlecap/main.rs:389]: do we need a lock here? this prevents us from processing any messages until it's acquired, what's it fully protecting here?

File: bottlecap/src/bin/bottlecap/main.rs (hunk 7 above)
Comment (astuyve) [reply bottlecap/src/bin/bottlecap/main.rs:420]: it'd be good to push this out somewhere which can do the work asynchronously, you're blocking the loop right now

**Review by astuyve — APPROVED** [review]

File: bottlecap/src/lifecycle/invocation/context.rs (hunk 2 above)
Comment (duncanista) [reply bottlecap/src/lifecycle/invocation/context.rs:31]: When a `PlatformRuntime` done hits, we effectively remove the `Context` in question, is this what you mean?

What do you mean by `drop invocation context which we haven't used for some reason yet`?

File: bottlecap/src/bin/bottlecap/main.rs (hunk 8 above)
Comment (duncanista) [reply bottlecap/src/bin/bottlecap/main.rs:389]: Looking what improvements I can make here

Comment (duncanista) [reply bottlecap/src/bin/bottlecap/main.rs:389]: As per our conversation, will move the lock out of this in another PR when the time comes, will add a ticket for this

**Pull request merged**